### PR TITLE
Zqs 1240/vqe api

### PR DESCRIPTION
**Pull request by Boniface316**
## Description

Include description of feature this PR introduces or a bug that it fixes. Include the following information:

This feature provides higher level API for VQE and simplifies user experience. 

## Please verify that you have completed the following steps

- [X] I have self-reviewed my code.
- [X] I have included test cases validating introduced feature/fix.

### CODE REVIEW

File: src/orquestra/vqa/algorithms/vqe.py [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+        cls,
+        hamiltonian: PauliRepresentation,
+        ansatz: Ansatz,
+        use_exact_expectation_values: bool = True,
+        grouping: str = "greedy",
+        shots_allocation: str = "proportional",
+        n_shots: Optional[int] = None,
+    ) -> "VQE":
+        """Creates a VQE object with some default settings:
+        - optimizer: L-BFGS-B optimizer (scipy implementation)
```
Comment (mstechly) [inline src/orquestra/vqa/algorithms/vqe.py:79]: There should be a newline in 79 and this line should be in 80

File: src/orquestra/vqa/algorithms/vqe.py [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+        Args:
+            hamiltonian: Cost Hamiltonian defining the problem.
+            use_exact_expectation_values: A flag indicating whether to use exact
+                calculation of the expectation values. This is possible only when
+                running on a simulator. Defaults to True.
+            grouping: name of the grouping function provided as a string. It only
+            accepts "greedy" and "individual" as an argument.
+            shots_allocation: name of the shots allocation function provided as a
+            string. It only accepts "proportional" and "individual" as an argument.
+            n_shots: If non-exact method for calculating expectation values is used,
```
Comment (mstechly) [inline src/orquestra/vqa/algorithms/vqe.py:97]: I would rephrase to:
"Specifies number of shots to be used for a given shots allocation method. If exact_expectation_values is true, it should be equal to `None`.

File: src/orquestra/vqa/algorithms/vqe.py [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+            values through averaging the results of measurements.
+
+        These can be later replaced using one of the `replace_*` methods.
+
+        Args:
+            hamiltonian: Cost Hamiltonian defining the problem.
+            use_exact_expectation_values: A flag indicating whether to use exact
+                calculation of the expectation values. This is possible only when
+                running on a simulator. Defaults to True.
+            grouping: name of the grouping function provided as a string. It only
```
Comment (mstechly) [inline src/orquestra/vqa/algorithms/vqe.py:93]: We should also allow for no grouping – cause for exact expectation values you don't need grouping, everything is co-measurable 🤔

File: src/orquestra/vqa/algorithms/vqe.py [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+    def get_circuit(self, params: np.ndarray) -> Circuit:
+        """Returns a circuit associated with give VQE instance.
+
+        Args:
+            params: ansatz parameters.
+        """
+        return self.ansatz.get_executable_circuit(params)
+
+
+def _get_grouping(grouping) -> EstimationPreprocessor:
```
Comment (mstechly) [inline src/orquestra/vqa/algorithms/vqe.py:296]: No typing here :)
I think it should be `Optional[grouping]`?

File: src/orquestra/vqa/algorithms/vqe.py (hunk 3 above)
Comment (mstechly) [reply src/orquestra/vqa/algorithms/vqe.py:93]: Also, this would mean that we should change type to `Optional[str]="greedy"` 🤔

Comment (mstechly) [reply src/orquestra/vqa/algorithms/vqe.py:93]: But in case `use_exact_expectation_values` is true, `greedy` should be overriden to `None` (as it is kind of anyway by the if statement in line 269 I think) , but we should throw a warning on top of that.

File: tests/orquestra/vqa/algorithms/vqe_test.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+from orquestra.vqa.grouping import group_greedily, group_individually
+from orquestra.vqa.shot_allocation import (
+    allocate_shots_proportionally,
+    allocate_shots_uniformly,
+)
+
+
+@pytest.fixture()
+def hamiltonian():
+    return PauliTerm("Z0") + PauliTerm("Z1")
```
Comment (mstechly) [inline tests/orquestra/vqa/algorithms/vqe_test.py:20]: For VQE you should use non-ising Hamiltonian.

File: tests/orquestra/vqa/algorithms/vqe_test.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+@pytest.fixture()
+def vqe_object(hamiltonian, ansatz):
+    return VQE.default(hamiltonian=hamiltonian, ansatz=ansatz)
+
+
+class TestDefaultVQE:
+    def test_default_optimizer_is_lbfgsb(self, vqe_object):
+        assert vqe_object.optimizer.method == "L-BFGS-B"
+
+    def test_default_estimation_init_default_value(self, vqe_object):
```
Comment (mstechly) [inline tests/orquestra/vqa/algorithms/vqe_test.py:52]: Rename to `test_default_estimation_method_is_...`? 
Same in the following.
I think you followed a pattern from a slightly outdated version of QAOA, check out how it's done on `main` now (I might be wrong though!)

File: tests/orquestra/vqa/algorithms/vqe_test.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+        assert np.isclose(
+            cost_function(np.zeros(vqe_object.ansatz.number_of_params)), 0
+        )
+
+    def test_get_circuit(self, vqe_object):
+        params = np.random.random(vqe_object.ansatz.number_of_params)
+        circuit = vqe_object.get_circuit(params)
+        assert circuit.n_qubits == vqe_object.ansatz.number_of_qubits
+
+    def test_replace_grouping(self, vqe_object):
```
Comment (mstechly) [inline tests/orquestra/vqa/algorithms/vqe_test.py:172]: I would move these tests up, so all the "replace tests` are in the same place.

**Review by mstechly — CHANGES_REQUESTED** [review]

File: tests/orquestra/vqa/algorithms/vqe_test.py (hunk 6 above)
Comment (Boniface316) [reply tests/orquestra/vqa/algorithms/vqe_test.py:52]: Its the same in the main

Comment (mstechly) [reply tests/orquestra/vqa/algorithms/vqe_test.py:52]: Well..
Could you refactor it in both QAOA and VQA to have better naming then 😅 ?

File: src/orquestra/vqa/algorithms/vqe.py (hunk 3 above)
Comment (Boniface316) [reply src/orquestra/vqa/algorithms/vqe.py:94]: Done

**Review by mstechly — APPROVED** [review]

**Review by mstechly — APPROVED** [review]
